### DISCUSSION TITLE
Invoking promises in @xstate/fsm

### DISCUSSION THREAD

**Discussion by jamesopstad** [statelyai/xstate#1232] [General] [answered]
Hi. Would it greatly increase the size of `@xstate/fsm` to add support for invoking promises? I'm working on a project where the bundle size needs to be kept very small so I'm trying to use this rather than the full xstate. I imagine adding support for invoking observables and other machines would add a lot of code but promises (and perhaps callbacks) may be reasonably small and are often needed.

**Comment by davidkpiano** [comment, marked as answer]
V5 of XState will be more modular/smaller. `@xstate/fsm` is kept intentionally small, so new features will likely not be added.

What you can do is execute an action that calls a promise, and when the promise is fulfilled, manually `send(...)` back an event to the machine:

```js
const fetchMachine = createMachine({
  initial: 'idle',
  states: {
    idle: {/* ... */},
    fetching: {
      entry: 'fetchFromAPI'
      on: {
        RESOLVE: 'success'
      }
    },
    success: {/* ... */}
  }
}, {
  actions: {
    fetchFromApi() {
      fetch(/* ... */).then(data => {
        service.send({ type: 'RESOLVE', data });
      });
    }
  }
});

const service = interpret(machine).start();
```

**Reply by jamesopstad** [reply]
That's an interesting solution. Thanks for the quick response. Maybe one day `xstate` will be tree shakable to the extent that `@xstate/fsm` is no longer necessary.

**Reply by davidkpiano** [reply]
Re: tree shaking, that's the plan! cc. @Andarist can give more information.
